src/middleware/subscribeWithSelector.ts

```diff
@@ -1,43 +1,62 @@
 import {
-  EqualityChecker,
-  GetState,
-  SetState,
   State,
+  StateCreator,
   StateListener,
-  StateSelector,
-  StateSliceListener,
   StoreApi,
+  StoreMutatorIdentifier,
   Subscribe,
 } from '../vanilla'
 
-export type StoreApiWithSubscribeWithSelector<T extends State> = StoreApi<T> & {
-  subscribe: Subscribe<T> & {
-    (listener: StateListener<T>): () => void
-    <StateSlice>(
-      selector: StateSelector<T, StateSlice>,
-      listener: StateSliceListener<StateSlice>,
-      options?: {
-        equalityFn?: EqualityChecker<StateSlice>
-        fireImmediately?: boolean
-      }
-    ): () => void
+type SubscribeWithSelector = <
+  T extends State,
+  Mps extends [StoreMutatorIdentifier, unknown][] = [],
+  Mcs extends [StoreMutatorIdentifier, unknown][] = []
+>(
+  initializer: StateCreator<
+    T,
+    [...Mps, ['zustand/subscribeWithSelector', never]],
+    Mcs
+  >
+) => StateCreator<T, Mps, [['zustand/subscribeWithSelector', never], ...Mcs]>
+
+declare module '../vanilla' {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  interface StoreMutators<S, A> {
+    ['zustand/subscribeWithSelector']: WithSelectorSubscribe<S>
   }
 }
 
-export const subscribeWithSelector =
-  <
-    S extends State,
-    CustomSetState extends SetState<S>,
-    CustomGetState extends GetState<S>,
-    CustomStoreApi extends StoreApi<S>
-  >(
-    fn: (set: CustomSetState, get: CustomGetState, api: CustomStoreApi) => S
-  ) =>
-  (
-    set: CustomSetState,
-    get: CustomGetState,
-    api: CustomStoreApi & StoreApiWithSubscribeWithSelector<S>
-  ): S => {
+type WithSelectorSubscribe<S> = S extends { getState: () => infer T }
+  ? S & StoreSubscribeWithSelector<Extract<T, State>>
+  : never
+
+interface StoreSubscribeWithSelector<T extends State> {
+  subscribe: <U>(
+    selector: (state: T) => U,
+    listener: (selectedState: U, previousSelectedState: U) => void,
+    options?: {
+      equalityFn?: (a: U, b: U) => boolean
+      fireImmediately?: boolean
+    }
+  ) => () => void
+}
+
+export type StoreApiWithSubscribeWithSelector<T extends State> =
```

Comment:
Do we need to export this type? Does a user need to use this for explicit typing?

Reply:
Nope, I'd prefer this to not be exported. But I kept it exported (it's already exported in v3) to avoid breaking changes.

We can deprecated this, in favor of writing `Mutate<Store<T>, ["zustsnd/subscribeWithSelector"]>` instead.

Reply:
> We can't give migration path in v3.7.0 unfortunately.

We can give a partial migration path by having `Mutate` (just `Mutate` no changes in `create` etc) in v3.7.0 itself and deprecate the old type(s). I'll open a PR in a few days if that's okay.

Reply:
Did it, see #772 to know what I mean